### PR TITLE
Add TypeScript to @truffle/compile-solidity and convert compilerSupplier module

### PR DESCRIPTION
**Pull request by gnidan**
(This follows #4222)

This PR adds TypeScript to @truffle/compile-solidity and sets it up as the `prepare` step. TS is configured to `allowJs` here, so as to leave most of the code in JS, but this PR does convert the `compilerSupplier/` module into TS.

Couple notes on this:

- There's odd behavior when doing a combination of `allowJs: true, declaration: true`, where if the `outDir` is inside the `rootDir`, then TS will generate declaration files for JS sources, then subsequently fail to compile with a TS5055 error ("Cannot write ... because it will overwrite input file"). Best I can figure, TS attempts to read `dist/<file>.d.ts` as input and also write it as output. **As a result**, I've set `rootDir` to `src/` rather than our conventional `./`. We'll want to change this in the future when we switch the entire package to TS and can turn off `allowJs`.
- A bunch of other Truffle packages do deep imports into some part of compile-solidity. These paths will need to include `dist/` from now on. This breaks the interface, naturally.
- All the TS files in `compilerSupplier/` now provide named exports only; this also is breaking.

### CODE REVIEW

File: .github/workflows/nodejs.yml
Comment (gnidan) [inline .github/workflows/nodejs.yml:10]: (Will need to take this out before merge)

File: packages/compile-solidity/package.json
Comment (cds-amal) [inline packages/compile-solidity/package.json:51]: Note, [ttypescript](https://github.com/cevek/ttypescript/issues/122) maintenance status is a concern

File: packages/compile-solidity/src/compilerSupplier/loadingStrategies/VersionRange.ts
Comment (cds-amal) [inline packages/compile-solidity/src/compilerSupplier/loadingStrategies/VersionRange.ts:84]: What does `solcVersion is string` do here? How is it different from `.filter((solcVersion: string) => !!solcVersion)`

**Review by cds-amal — APPROVED** [review]
Looks good to me. Though it's probably safer to get more eyes on the TS

File: packages/compile-solidity/package.json
Comment (gnidan) [reply packages/compile-solidity/package.json:51]: Understood, but I'm hoping we can use in this PR and figure out a solution across the board.

File: packages/compile-solidity/src/compilerSupplier/loadingStrategies/VersionRange.ts
Comment (gnidan) [reply packages/compile-solidity/src/compilerSupplier/loadingStrategies/VersionRange.ts:84]: It's a [User-defined type guard](https://www.typescriptlang.org/docs/handbook/advanced-types.html#user-defined-type-guards); TS knows that the `boolean` returned by that function informs the type of `solcVersion`

Comment (gnidan) [reply packages/compile-solidity/src/compilerSupplier/loadingStrategies/VersionRange.ts:84]: (Since the `.map()` operation will possibly return `undefined`, returning a type predicate like that means that TS won't think the type of `satisfyingVersions` is `string | undefined`)

Comment (cds-amal) [reply packages/compile-solidity/src/compilerSupplier/loadingStrategies/VersionRange.ts:84]: TIL. Interesting syntax. Thanks!

File: packages/compile-solidity/src/compilerSupplier/normalizeSolcVersion.ts
Comment (haltman-at) [inline packages/compile-solidity/src/compilerSupplier/normalizeSolcVersion.ts:5]: Exported functions should really have type annotations, I think.

File: packages/compile-solidity/src/compilerSupplier/observeListeners.ts
Comment (haltman-at) [inline packages/compile-solidity/src/compilerSupplier/observeListeners.ts:27]: OK, I pretty much ignored the listener stuff on the last PR, but this `@ts-expect-error` has made me curious.  I'm a little confused here because:

1. How are we using `process` without importing it?
2. What's this `process.listeners()` function, given that I can't find it documented in the Node API docs?
3. What type does it actually take (that apparently includes strings), what type is it typed as taking (that apparently does not include strings), and why is `@ts-expect-error` better here than some sort of conversion or coercion?

File: packages/compile-solidity/src/compilerSupplier/rangeUtils.ts
Comment (haltman-at) [inline packages/compile-solidity/src/compilerSupplier/rangeUtils.ts:11]: I think these functions should get type annotations, since they're exported.

File: packages/core/lib/testing/SolidityTest.js
Comment (haltman-at) [inline packages/core/lib/testing/SolidityTest.js:6]: Geez, I hope that goes away in an upcoming PR... maybe mark this?

File: packages/core/lib/testing/Test.js
Comment (haltman-at) [inline packages/core/lib/testing/Test.js:13]: Similarly with these two.

File: packages/resolver/lib/sources/truffle/Deployed.ts
Comment (haltman-at) [inline packages/resolver/lib/sources/truffle/Deployed.ts:3]: ...isn't there some way to make this work without `/dist/`?  Is that coming in one of the later PRs?

**Review by haltman-at — COMMENTED** [review]
Mostly seems fine, a few things I'm confused about.

File: packages/compile-solidity/src/compilerSupplier/index.ts
Comment (gnidan) [reply packages/compile-solidity/src/compilerSupplier/index.ts:10]: I'm fine switching to `any` if you think it's better; I generally default to using `unknown`, but I don't have an argument for it here.

File: packages/compile-solidity/src/compilerSupplier/loadingStrategies/Docker.ts
Comment (gnidan) [reply packages/compile-solidity/src/compilerSupplier/loadingStrategies/Docker.ts:89]: I'm pretty sure Husky ran for me at some point while doing this work?

Guess we'll see...

File: packages/compile-solidity/src/compilerSupplier/index.ts
Comment (haltman-at) [reply packages/compile-solidity/src/compilerSupplier/index.ts]: OK, but what about the rest?

File: packages/compile-solidity/src/compilerSupplier/loadingStrategies/Docker.ts
Comment (haltman-at) [reply packages/compile-solidity/src/compilerSupplier/loadingStrategies/Docker.ts:89]: I guess?

File: packages/compile-solidity/src/compilerSupplier/observeListeners.ts
Comment (gnidan) [reply packages/compile-solidity/src/compilerSupplier/observeListeners.ts:27]: > 1. How are we using `process` without importing it?

It's a Node.js global.

> 2. What's this `process.listeners()` function, given that I can't find it documented in the Node API docs?

Oh jeez, good question. I don't know where I found this; I added it originally back in #4092. TS seems to know it? Weird.

> 3. What type does it actually take (that apparently includes strings), what type is it typed as taking (that apparently does not include strings), and why is `@ts-expect-error` better here than some sort of conversion or coercion?

Hm, it looks like I can cast this `eventName as NodeJS.Signals` and it compiles... lately I've taken to using `@ts-ignore` or `@ts-expect-error` instead of casting, since I don't (as a habit) want to muck with the actual types of things when I can just tell the compiler not to care about something.

Here it doesn't matter either way, but in some situations, `as` can remove all kinds of type information that you might want elsewhere.

File: packages/compile-solidity/src/compilerSupplier/loadingStrategies/Docker.ts
Comment (gnidan) [reply packages/compile-solidity/src/compilerSupplier/loadingStrategies/Docker.ts]: Did so

File: packages/compile-solidity/src/compilerSupplier/index.ts
Comment (gnidan) [reply packages/compile-solidity/src/compilerSupplier/index.ts]: Fixed those. Thanks for catching them; I meant to go back and add real types.

File: packages/compile-solidity/src/compilerSupplier/loadingStrategies/Docker.ts
Comment (gnidan) [reply packages/compile-solidity/src/compilerSupplier/loadingStrategies/Docker.ts:89]: Lemme know if you want me to play it safe and change it :)

File: packages/compile-solidity/src/compilerSupplier/loadingStrategies/VersionRange.ts
Comment (gnidan) [reply packages/compile-solidity/src/compilerSupplier/loadingStrategies/VersionRange.ts]: (Added a comment here)

File: packages/compile-solidity/src/compilerSupplier/observeListeners.ts
Comment (haltman-at) [reply packages/compile-solidity/src/compilerSupplier/observeListeners.ts:27]: Crazy!  Yeah let's go with that rather than using `@ts-expect-error`.

File: packages/core/lib/testing/SolidityTest.js
Comment (gnidan) [reply packages/core/lib/testing/SolidityTest.js:6]: It doesn't go away.

I don't have a good answer to this: in db and codec, etc., the approach we've settled on is to export everything in one module, subdivided into namespaces. Unfortunately, this means you can't do any tree-shaking; all the modules everywhere end up imported all at once, resulting in a large module size in memory and risk of incompatibility with things like `fs`.

I've left these here for now because I didn't want to go ahead and follow this prior approach here, knowing now the pretty severe downsides. My conclusion was that this deep-package import could really be considered internal to Truffle (users of the package very likely won't need to do this), and so it seemed acceptable for the time being.

Would love to discuss this though.

File: packages/compile-solidity/src/compilerSupplier/observeListeners.ts
Comment (haltman-at) [reply packages/compile-solidity/src/compilerSupplier/observeListeners.ts:27]: (Although I still don't understand your claim that casting/coercing loses type info, since it only loses info in that particular expression... it doesn't lose info when you use that variable without a cast or coercion...)

File: packages/core/lib/testing/Test.js
Comment (gnidan) [reply packages/core/lib/testing/Test.js:13]: (Resolving this to move discussion to [other comment thread](https://github.com/trufflesuite/truffle/pull/4223#discussion_r688190965))

File: packages/resolver/lib/sources/truffle/Deployed.ts
Comment (gnidan) [reply packages/resolver/lib/sources/truffle/Deployed.ts:3]: 😢 

(Resolving this to move discussion to [other comment thread](https://github.com/trufflesuite/truffle/pull/4223#discussion_r688190965))

File: packages/compile-solidity/src/compilerSupplier/loadingStrategies/Docker.ts
Comment (haltman-at) [reply packages/compile-solidity/src/compilerSupplier/loadingStrategies/Docker.ts:89]: Eh let's leave it in for now and we can change it if it causes problems I guess.

File: packages/compile-solidity/src/compilerSupplier/observeListeners.ts
Comment (gnidan) [reply packages/compile-solidity/src/compilerSupplier/observeListeners.ts:27]: > (Although I still don't understand your claim that casting/coercing loses type info, since it only loses info in that particular expression... it doesn't lose info when you use that variable without a cast or coercion...)

Well, if your expression gets passed to some function that doesn't have an explicit type annotation, that function will receive the casted type. This has bitten me enough times that I now avoid `as` like the plague.

Comment (gnidan) [reply packages/compile-solidity/src/compilerSupplier/observeListeners.ts]: But here, have an `as` :)

File: packages/core/lib/testing/SolidityTest.js
Comment (haltman-at) [reply packages/core/lib/testing/SolidityTest.js:6]: Blech. :-/  Well, if you think this is the approach to take, I'll leave it to you.

File: packages/compile-solidity/src/compilerSupplier/index.ts
Comment (haltman-at) [reply packages/compile-solidity/src/compilerSupplier/index.ts]: Hey, seems you missed a few here?  You got the ones inside `strategyOptions` but not outside.

**Review by haltman-at — COMMENTED** [review]
[My review is that you should fix those few types that you missed that I mentioned above. :) ]

File: packages/compile-solidity/src/compilerSupplier/observeListeners.ts
Comment (haltman-at) [reply packages/compile-solidity/src/compilerSupplier/observeListeners.ts]: > Well, if your expression gets passed to some function that doesn't have an explicit type annotation, that function will receive the casted type.

I am so confused as to why you would ever do this!  If the function doesn't have an explicit type annotation (and why doesn't it...?) then why are you even casting what you pass in?  But I think if anything the lesson here is to put type annotations on your functions...

File: packages/compile-solidity/src/compilerSupplier/index.ts
Comment (gnidan) [reply packages/compile-solidity/src/compilerSupplier/index.ts]: Fixed those!

**Review by haltman-at — APPROVED** [review]
Alright, looks good!

**Comment by gnidan** [comment]
Thanks @haltman-at and @cds-amal!

**Pull request merged**
